notebooks/integrations/openai/function-calling.ipynb

```diff
@@ -0,0 +1,502 @@
+    "import json\n",
+    "import requests"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "578a47e0-f9d7-49e5-a2b6-c417378392c3",
+   "metadata": {},
+   "source": [
+    "## Add Credentials"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "d052237f-5874-4f25-b290-5a5f87006536",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "OPENAI_API_KEY = getpass(\"OpenAI API Key:\")\n",
+    "client = OpenAI(\n",
+    "    api_key=OPENAI_API_KEY,\n",
+    ")\n",
+    "GPT_MODEL = input(\"GPT Model:\")\n",
+    "\n",
+    "ES_API_KEY = getpass(\"Elastic API Key:\")\n",
+    "ES_ENDPOINT = input(\"Elasticsearch Endpoint:\")\n",
+    "ES_INDEX = \"kibana_sample_data_flights\"\n",
+    "\n",
+    "WU_API_KEY = getpass(\"Weather Union API Key:\")\n",
+    "WU_API_ENDPOINT = input(\"Weather Union API Endpoint:\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a255228d-7994-4bbe-ae53-bb574ea320f8",
+   "metadata": {},
+   "source": [
+    "## Functions to get data from Elasticsearch"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "06d09f3e-38b3-43c5-a156-5b3096f45d99",
+   "metadata": {},
+   "source": [
+    "### Get Index mapping"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "9c67afad-c0a0-4bab-98c8-5f8da582336a",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def get_index_mapping():\n",
+    "\n",
+    "    url = f\"\"\"{ES_ENDPOINT}/{ES_INDEX}/_mappings\"\"\"\n",
+    "\n",
+    "    headers = {\n",
+    "        \"Content-type\": \"application/json\",\n",
+    "        \"Authorization\": f\"\"\"ApiKey {ES_API_KEY}\"\"\",\n",
+    "    }\n",
+    "\n",
+    "    resp = requests.request(\"GET\", url, headers=headers)\n",
+    "    resp = json.loads(resp.text)\n",
+    "    mapping = json.dumps(resp, indent=4)\n",
+    "\n",
+    "    return mapping"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c37bf4b0-fb96-4cff-9ddc-9aef781e30bb",
+   "metadata": {},
+   "source": [
+    "### Get reference document"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "0db00590-e417-4751-a0c6-8e5a2f924fa1",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def get_ref_document():\n",
+    "\n",
+    "    url = f\"\"\"{ES_ENDPOINT}/{ES_INDEX}/_search?size=1\"\"\"\n",
+    "\n",
+    "    headers = {\n",
+    "        \"Content-type\": \"application/json\",\n",
+    "        \"Authorization\": f\"\"\"ApiKey {ES_API_KEY}\"\"\",\n",
+    "    }\n",
+    "\n",
+    "    resp = requests.request(\"GET\", url, headers=headers)\n",
+    "    resp = json.loads(resp.text)\n",
+    "    json_resp = json.dumps(resp[\"hits\"][\"hits\"][0], indent=4)\n",
+    "\n",
+    "    return json_resp"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "6dc83930-89c0-465b-9002-e6aa0d3f5e6f",
+   "metadata": {},
+   "source": [
+    "### Generate Elasticsearch Query DSL based on user query"
```

Comment:
Worth looking into few-shot prompting so you can provide examples on the types of queries you want it to return. This should increase consistency.

Reply:
Hi Joe, I think adding this suggestion as side note would be better. Indeed its great suggestion but performing NL query is one of the part but focus is function calling. As I tested it working for most of the basic query i.e search, aggs.

Reply:
sure! Likely it will work fine as is and yes mention in blog that we are relying on the model's own knowledge to generate the queries but we can increase reliability with few shot for more complex queries.

Reply:
Good input. Thanks Joe